### DISCUSSION TITLE
Analysis Latency

### DISCUSSION THREAD

**Discussion by dbrianhendrix** [mcguirepr89/BirdNET-Pi#249] [Q&A]
For whatever reason I can't figure out, BirdNET-Pi has stopped detecting bird songs. It is running and I can see it analyzing and there are clearly high quality recordings of bird song. However, for the last 4 hours, nothing has been detected. I can hear the birds clearly as well. The detections stopped around 1pm CST and it's after 5pm now.

<img width="1355" alt="Screen Shot 2022-05-06 at 5 09 49 PM" src="https://user-images.githubusercontent.com/52499277/167222510-4d3c71fa-33da-4cdd-b79c-de87bbabac4d.png">
.

**Comment by mcguirepr89** [comment]
Hi, Brian! That sucks!!

Would you mind sending me a screenshot of the "View Log" page so I can try to see what is going on?

**Reply by dbrianhendrix** [reply]
Not sure if I should done "reply" or another message below.

**Comment by dbrianhendrix** [comment]
Well, here's a snapshot -- it was very actively scrolling.

<img width="1629" alt="Screen Shot 2022-05-06 at 5 23 59 PM" src="https://user-images.githubusercontent.com/52499277/167223678-66c9a446-0d8b-4cdd-8711-1770b31ec932.png">
.

**Reply by dbrianhendrix** [reply]
Wait! Are those time stamps right? Is it processing things from 2pm CST?

**Comment by dbrianhendrix** [comment]
Okay, I'm now seeing things happen I missed before. The 3pm "slot" on the "Overview" page is now entering detections. So my assumption here is that BirdNET-Pi is quite a bit behind in analyzing WAV files which explains why the "Last Hour" is saying "0".

Are there some tweaks to configuration I need to make to speed this up?

**Reply by ehpersonal38** [reply]
BirdNET-Pi will go back and analyze all recordings it missed (in /home/{user}/BirdSongs) during a period of downtime, so that's likely what you're seeing-- the system catching back up. Not sure why it went down in the first place though. 

Unfortunately you'll just have to wait - an analysis takes about 1.5 seconds, so multiply that by the number of files in /home/{user}/BirdSongs and you have a rough estimate of how long it'll take. You won't lose any recordings though, current recordings will just be added to the back of the queue.

**Reply by mcguirepr89** [reply]
Yep it looks like it is super backed up on analysis. I'm not super sure what is causing this for you :thinking:

Could you take a look at "Tools" > "System Info" and let me know what these blocks look like for you?
![image](https://user-images.githubusercontent.com/60325264/167225286-fcc8a59f-c089-49c1-82d8-896aa894d5fd.png)
![image](https://user-images.githubusercontent.com/60325264/167225320-ed724f28-521a-4313-ae1b-81b12b809d26.png)
and
![image](https://user-images.githubusercontent.com/60325264/167225334-eb067de0-e12e-47e1-be9e-d5aeef185dbd.png)

and finally, just this section of the "Viewer" box:
```
..................................`vcgencmd stats`..............................
--
throttled=0x0
Binary: 0
```


This seems like a hardware issue hitting some kind of limit. Could be:
- Throttling due to overheating in the middle of the day?
- RAM that's too small and needs a little help from a larger swap

Or, if it's not that, we can troubleshoot further if you feel comfortable entering commands in the terminal and sending me their output, we can try to see if there is something else going on.

**Reply by mcguirepr89** [reply]
Upon further review, I think the real culprit is the time it takes to `POST` to BirdWeather.com -- you were posting a TON of detections today!

**Reply by dbrianhendrix** [reply]
Here is the information requested:

<img width="1206" alt="Screen Shot 2022-05-06 at 10 31 35 PM" src="https://user-images.githubusercontent.com/52499277/167236534-e72398b5-21eb-432e-8497-a6cac3ae5713.png">
<img width="1224" alt="Screen Shot 2022-05-06 at 10 31 04 PM" src="https://user-images.githubusercontent.com/52499277/167236537-7dc58d18-5cec-4207-9b03-e8e109e9f9c6.png">
<img width="1162" alt="Screen Shot 2022-05-06 at 10 33 12 PM" src="https://user-images.githubusercontent.com/52499277/167236532-16e756f1-daf0-499c-860f-02a1e4e1963a.png">
<img width="1156" alt="Screen Shot 2022-05-06 at 10 32 12 PM" src="https://user-images.githubusercontent.com/52499277/167236533-cc8ca152-edb0-4e46-bb97-0087e9e7517d.png">

**Reply by mcguirepr89** [reply]
Thanks, BH -- that all looks reasonable and does not indicate any hardware issues from what I can tell. It does seem that this delay is due to the `POST`ing to BirdWeather.com.

For context, there is a section in the `server.py` script that handles the analysis wherein the timestamp for the recording is used for the `POST` -- Tim had to add this precisely to address latency from bird-abundant hours. :/

If you want to ensure your system won't get behind again, I'm afraid the only option would be to discontinue using BirdWeather.com, which would suck as you would no longer be sharing your data with that platform :/

I'm so sorry that this is happening for you! This is a new dilemma, as you're the first user whose installation is approaching being _too_ productive to use BirdWeather.com! A novel and cool issue to encounter :thinking:

I'm going to continue to think on this and will let you know what other options I'm able to suggest.

My very best regards,
Patrick

PS
I hope your dinner date was a lovely one :) :heart:

**Reply by ehpersonal38** [reply]
Good morning Patrick, 

Could asynchronous POST requests be an option here?

-Ethan

**Reply by mcguirepr89** [reply]
Good morning!

I have a really limited knowledge about those types of things, but based on [a _very_ cursory look at the `POST` options](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/Synchronous_and_Asynchronous_Requests), it does seem that asynchronous `POST` requests are preferred over synchronous. To be honest, at this point I'm not sure of how to configure one over the other, but will look into that.

Thank you for the great suggestion!

**Reply by ehpersonal38** [reply]
Great! It can be a bit of a headache since I'm not sure you can call async functions from within synchronous code. There must be some way to do it though...

**Reply by dbrianhendrix** [reply]
New problem that may be related. It appears now system stuck at some sort of housekeeping on some directories. Nothing is being collected and analyzed now.

<img width="1630" alt="Screen Shot 2022-05-07 at 9 23 12 AM" src="https://user-images.githubusercontent.com/52499277/167258485-09e7f032-7cce-43bb-abbf-b2980417a53f.png">

**Reply by mcguirepr89** [reply]
Hi, Brian -- do me a favor and do another update: "Tools" > "System Controls" > "Update"

**Reply by ehpersonal38** [reply]
I suppose we should make an issue titled "lsof: command not found" in anticipation of users updating?

**Reply by mcguirepr89** [reply]
Hahah probably

**Reply by mcguirepr89** [reply]
@dbrianhendrix 

The update may take a minute as it is installing a new command. If that doesn't fully clear things up, I recommend going to "Tools" > "Services" > "Restart ALL Services"

You'll lose the data from today, but your system will be in a better place after that.

Alternatively, if you don't want to lose any data from today, you can try:
"Tools" > "Services" > "Stop Core Services"
wait 5 minutes
"Tools" > "Services" > "Restart" BirdNET Analysis Server
"Tools" > "Services" > "Restart" BirdNET Analysis Client
"Tools" > "Services" > "Restart" Recording Service
"Tools" > "Services" > "Restart" Extraction

I'm hopeful, though, that the second update will address the issue. :crossed_fingers:

**Reply by dbrianhendrix** [reply]
I decided to just reboot. It's now processing audio data from 11pm last night which must was where it was "stuck". I also did some research (I am a Senior Linux Engineer) and notice that swap is not big enough. The swap was only 100MB and the general rule is to have twice the RAM in swap. The 100MB is now 8GB.

root@birdnetpi:# swapon -s
Filename                                Type            Size    Used    Priority
/var/swap                               file            8388604 0       -2

root@birdnetpi:~# free -h
               total        used        free      shared  buff/cache   available
Mem:           3.7Gi       627Mi       2.0Gi       3.0Mi       1.1Gi       3.0Gi
Swap:          8.0Gi          0B       8.0Gi

I honestly don't believe swap was an issue but it definitely shouldn't be now.

I guess we can turn off POST to app.birdweather.com but that would be unfortunate. So far, it appears this new error didn't stop the recording of bird song as I'm seeing analysis happening at 1am today.

<img width="1614" alt="Screen Shot 2022-05-07 at 9 38 08 AM" src="https://user-images.githubusercontent.com/52499277/167258974-ff7f79f6-07bb-4440-81e9-1e0316a48ca9.png">

**Reply by mcguirepr89** [reply]
Great to know you have a Linux background!! You'll start noticing many departures from Linux sysadmin best praxis in favor of some Raspberry Pi-isms, but that's mostly just to say that BirdNET-Pi leverages `systemd` and the `NOPASSWD` sudo policy for the default user. The effect is basically administering the system completely through the default user.

You're spot on regarding best practice for the swap space, though I've followed that basic rule of thumb, and found my system never utilized more than 100M of the swap anyways :/ so I've discontinued that step of the installation (I used to use the zRAM kernel module for swapping in the original version of BirdNET-Pi called "BirdNET-system" but found the CPU use not worth the speed increase in swap performance.

Truly I think the only thing going on is your system having a REALLY great mic that is picking up a LOT of data and the system is trying to POST it in a way that is really clogging up the system.

I'll try to work with Tim on this to see what we can do to mitigate the latency introduced in the `POST`s.

Also, as a Linux sysadmin, you'll appreciate that all services are `journal`ed and logs can be access with `journalctl` -- so if you need to see the log output from 5 minutes before 1am through 2am for all of the core servcies, something like:
```
journalctl --no-hostname -ubirdnet_analysis -ubirdnet_server -uextraction -ubirdnet_recording --since="00:55:00" --until="02:00:00"
```
will tell you a lot

**Reply by dbrianhendrix** [reply]
Update complete. I see the "lsof" unpacking and setting up at end of update.

**Reply by dbrianhendrix** [reply]
This is new!

<img width="1613" alt="Screen Shot 2022-05-07 at 10 26 38 AM" src="https://user-images.githubusercontent.com/52499277/167260913-bb70c275-78e0-487a-a32e-49fbb8bfc89a.png">

**Reply by mcguirepr89** [reply]
That means something is up with BirdWeather's API -- I'll let Tim know :+1:

**Reply by dbrianhendrix** [reply]
Should I turn off sending to the API for now and let system catch up? It's currently 6 hours behind.

**Reply by dbrianhendrix** [reply]
To turn off POSTing to API, do I just remove (blank) the "BirdWeather ID:" field under "Basic Settings"?

**Reply by dbrianhendrix** [reply]
Well, I tried that -- blanked the "BirdWeather ID:" field and saved. It stopped analyzing old WAV files and is now just doing current recordings. I put the ID back in the field and it's still not going back the missing 5 hours. Now it's just doing current.

**Reply by dbrianhendrix** [reply]
It appears the "Cannot POST right now" is not everyone. I'm seeing other sites saying they've received POSTs in the last few minutes. I'm wondering if he has "code" that's filtering me out currently.

**Reply by mcguirepr89** [reply]
@timsterc -- Any chance there is a rate limit on the API `POST`s?

**Reply by timsterc** [reply]
Server is *definitely* running slowly today!  There are no built-in limits (that I'm aware of), I think it's just the server getting slow.  Will probably have to wait until Sam is online again later tonight to get to the bottom of it.

**Reply by timsterc** [reply]
@dbrianhendrix - could you capture the POST message, including your BirdWeather ID and send to me.  You're welcome to email to me directly if you like.  As you've noted, I'm seeing POSTs from other BirdNET-Pi stations, so I'm wondering if this might be a BW ID issue.

**Reply by mcguirepr89** [reply]
I went ahead and removed the quotes for the BirdWeather ID so we'll see if that helps
https://github.com/mcguirepr89/BirdNET-Pi/commit/c3908a0435153de4076a0e93bb68a299bdb23906

**Reply by timsterc** [reply]
It looks like that did the trick - thanks @dbrianhendrix for catching this, and @mcguirepr89 for the fix!

In summary - if you're not seeing detections posting to BirdWeather - try an update.

**Comment by dbrianhendrix** [comment]
body{font-family:Helvetica,Arial;font-size:13px}I’m going to go ahead and do an update to make sure the new code works. I’m running now off just my edit.  D. Brian Hendrixveteran  |  mac guy  |  birder  |  coffee snobTennessee Naturalist, firstSenior Linux Engineer, ***@***.*** On May 8, 2022 at 11:28:10 AM, Tim Clark ***@***.***) wrote: 
It looks like that did the trick - thanks @dbrianhendrix for catching this, and @mcguirepr89 for the fix!
In summary - if you're not seeing detections posting to BirdWeather - try an update.

**Comment by mcguirepr89** [comment]
I'm seeing different results on GitHub for this conversation than what I have in my emails (from github) for this conversation -- it's making it a little tricky to follow

@dbrianhendrix It looks like your system is in better shape now that you guys found the quotations as the issue, is that correct?

**Reply by dbrianhendrix** [reply]
Absolutely! And THANKS! It's tricky to follow because I started to respond to Tim via email instead of including the github address. My fault!

**Reply by mcguirepr89** [reply]
Got it! Well I'm very glad your system is feeling better now! I'll try to grab the specific response on that email thread I wanted to address -- stay tuned (but don't hold your breath!!! I get easily distracted)
